File: clojure/rotational-cipher/src/rotational_cipher.clj
(ns rotational-cipher)

(def lowest-upper-ascii 65)
(def highest-upper-ascii 90)
(def lowest-lower-ascii 97)
(def highest-lower-ascii 122)

(defn alphabet-plus
  "If `a` + `b` is over boundary, wrap around the alphabet by subtracting by the
  count of characters, i.e. 26."
  [boundary a b]
  (let [combined (+ a b)]
    (if (> combined boundary)
      (- combined 26)
      combined)))

(defn rotate-ascii
  [ascii-val key]
  (cond
    (<= lowest-upper-ascii ascii-val highest-upper-ascii)
    (alphabet-plus highest-upper-ascii ascii-val key)

    (<= lowest-lower-ascii ascii-val highest-lower-ascii)
    (alphabet-plus highest-lower-ascii ascii-val key)

    :else ascii-val))

(defn rotate-char
  [input-char key]
  (char (rotate-ascii (int input-char) key)))

(defn rotate
  [input key]
  (apply str (map #(rotate-char % key) input)))

File: python/rotational-cipher/rotational_cipher.py
def rot_plus(boundary, char, key):
    combined = char + key
    if combined > boundary:
        return combined - 26
    return combined

def rotate_char(char, key):
    converted = ord(char)
    if ord("a") <= converted <= ord("z"):
        return chr(rot_plus(ord("z"), converted, key))
    if ord("A") <= converted <= ord("Z"):
        return chr(rot_plus(ord("Z"), converted, key))
    return char

def rotate(text, key):
    return "".join([rotate_char(char, key) for char in text])

def rotate_translate(text, key):
    "Use Python's <str>.translate method to rotate `text`. I like this a lot more than the ASCII code solution, especially the slicing by `key` to create the rotated character set. Taken from https://exercism.org/tracks/python/exercises/rotational-cipher/approaches/str-translate"
    alphabet = "abcdefghijklmnopqrstuvwxyz"
    translator = alphabet[key:] + alphabet[:key]
    return text.translate(str.maketrans(alphabet + alphabet.upper(), translator + translator.upper()))